Use systemd for cgroup2 split-tunneling

hulthe opened the pull request:
<!--
PR checklist. Does not need to be included in the submitted PR, but must be honored:

* [ ] The change is added to `CHANGELOG.md` under the `[Unreleased]` header.
* [ ] The change/commits follow the Mullvad coding guidelines: https://github.com/mullvad/coding-guidelines
* [ ] Automatic tests are added for the change, if relevant. All new features must have tests.
* [ ] The PR description should describe:
  * **What** this PR changes
  * **Why** this is wanted
  * If necessary, **how** it's implemented
  * How to **test** the change


👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋
  THIRD PARTY CONTRIBUTOR, PLEASE READ THIS
👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋👋

## Translations and localization

Do you want to contribute translations/localization to this app?
* If you want to correct an existing translation, please fill in this form instead of submitting
  a PR with changes to the PO/xml files: https://docs.google.com/forms/d/e/1FAIpQLSeEFRe0ojdl6QdHPp7Z9qIvdGTc1uSgbswQT6d-VRQ98GBO2w/viewform
* We can't accept translations to new languages from third party contributors.
-->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/mullvadvpn-app/9588)
<!-- Reviewable:end -->

hulthe committed:
Implement mullvad-exclude using systemd-run

Co-authored-by: David Lönnhager <david.l@mullvad.net>

dkwo commented:
Would this change imply that mullvad-exclude will only work with systemd?
is there any other solution that would not require that?
cc: @meator @MarkusPettersson98

hulthe commented:
@dkwo This change requires systemd, yes.

We will have a fallback to the old net_cls cgroup. This would apply to distros with old kernels, or without systemd. This fallback might be removed in the future though, since we don't really want to maintain support for deprecated kernel interfaces.

We're thinking about adding a fallback to some other non-systemd mechanism, but no decision yet.

dlon committed:
Intercept socket syscalls with seccomp in mullvad-exclude

This can be used with flatpak to "atomically" install our BPF hook in the new cgroup when the process is moved

Co-authored-by: Joakim Hulthe <joakim.hulthe@mullvad.net>

dlon committed:
Check libseccomp API version

dlon committed:
Automatically enable intercept in mullvad-exclude for flatpaks

Also add flags to override this behavior

dlon committed:
fixup! Automatically enable intercept in mullvad-exclude for flatpaks

hulthe committed:
Install seccomp dependencies in Dockerfile

dlon committed:
Fix CTRL-C killing the seccomp supervisor

dlon committed:
Spawn supervisor process safely

dlon committed:
fixup! Spawn supervisor process safely

Werkov commented:
@dkwo It boils down to arbitration over the cgroup tree. The implementation needs to know how to interact with it (how the tree looks like and how subtree delegation works to place own cgroups into in). There are no theoretical reasons tying it to systemd but practically, each environment needs adapted integration.

(The former situation with `nel_cls` is similar, if one focuses on `net_cls` as the single hierarchy -- it mostly worked because it wasn't used by anything else but if there were multiple programs trying to make use of it, some coordination becomes necessary.)

dkwo commented:
@Werkov Thanks for clarifying this!

waltmck commented:
I would just like to say that I am very excited to see this get merged! This will make split-tunneling daemons (like geoclue2 and mesh VPNs) much less janky.
